### DISCUSSION TITLE
How did vaultwarden bundle database with CRUD API in single docker image?

### DISCUSSION THREAD

**Discussion by winston0410** [dani-garcia/vaultwarden#2202] [General]
I am really curious how did vaultwarden did that with [their image](https://hub.docker.com/r/vaultwarden/server), as I always thought you need something like `docker-compose` to orchestrate multiple containers(the api and the database)

I have read the source for the [`Dockerfile`](https://github.com/dani-garcia/vaultwarden/blob/main/docker/amd64/Dockerfile), which doesn't define anything for the database, and points to `vaultwarden/web-vault`

And for the image of `vaultwarden/web-vault`, I can only find definition for [building the api](https://github.com/dani-garcia/bw_web_builds/blob/master/Dockerfile). So where did vaultwarden defined logic for databases?

**Comment by RealOrangeOne** [comment]
When using a single image, vaultwarden uses sqlite for the database, so there's no extra processes to bundle or run. The application _is_ the database engine (technically, but also not really). `vaultwarden/web-vault` is exactly that, just the web vault. `vaultwarden/vaultwarden` is the server-side component (which does the database comms)

**Reply by winston0410** [reply]
I see that make sense. In that case SQLite will be embedded, so one image will be fine. That's a good design decision there!

**Reply by RealOrangeOne** [reply]
sqlite is a pretty great technology, and yes because it's compiled in requires nothing extra, but it doesn't necessarily scale well. For multiple users, or just for some ease, you're often better off using something like an external postgres DB. Using sqlite by default is a fairly common design decision for applications which support multiple database backends, because it's the easiest to get started with.
